shapely/affinity.py

```diff
@@ -54,33 +54,22 @@ def affine_transform(geom, matrix):
             ndim = 3
             i = 1.0
             c = f = g = h = zoff = 0.0
-            matrix = a, b, c, d, e, f, g, h, i, xoff, yoff, zoff
     elif len(matrix) == 12:
         ndim = 3
         a, b, c, d, e, f, g, h, i, xoff, yoff, zoff = matrix
         if not geom.has_z:
             ndim = 2
-            matrix = a, b, d, e, xoff, yoff
     else:
         raise ValueError("'matrix' expects either 6 or 12 coefficients")
+    if ndim == 2:
+        A = np.array([[a, b], [d, e]], dtype=float)
+        off = np.array([xoff, yoff], dtype=float)
+    else:
+        A = np.array([[a, b, c], [d, e, f], [g, h, i]], dtype=float)
+        off = np.array([xoff, yoff, zoff], dtype=float)
 
     def _affine_coords(coords):
-        """Internal function to yield affine transform of coordinate tuples"""
-        x = coords[:, 0]
-        y = coords[:, 1]
-
-        if ndim == 2:
-            xp = a * x + b * y + xoff
-            yp = d * x + e * y + yoff
-            return np.hstack((np.atleast_2d(xp).T, np.atleast_2d(yp).T))
-        elif ndim == 3:
-            z = coords[:, 2]
-            xp = a * x + b * y + c * z + xoff
-            yp = d * x + e * y + f * z + yoff
-            zp = g * x + h * y + i * z + zoff
-            return np.hstack(
-                (np.atleast_2d(xp).T, np.atleast_2d(yp).T, np.atleast_2d(zp).T)
-            )
+        return np.dot(A, coords.T).T + off
```

Comment:
Would `np.dot(coords, A)` (without the double transpose) also work?

Reply:
that would raise (e.g.) "ValueError: shapes (2,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)". Transpose is an unchanged view of the original array, so it is a fast operation. I should also note that there are two alternative forms using matrix multiplcation:

1.  `np.matmul(A, coords.T).T + off`
2. `(A @ coords.T).T + off` for [PEP 465](https://peps.python.org/pep-0465/)